Reached heap limit when building since v4.2

### Describe the bug

Update: Looks like the sourcemap CLI option broke, no performance issue. See comments below.

---

Reopening this as it is still not solved. I don't know how to create a reproduction without sharing my whole codebase because I don't know what causes this. I'm open for suggestions though.

See comments below, looks like a performance issue introduced in 4.2.0.

---

I noticed since upgrading to 4.2 that my builds are failing on CI (self hosted runner for GH Actions, 4GB RAM). I'm running `vite build --mode testing --sourcemap false` with a React project. Since I'm not building source maps, I think it's unusual to reach the heap limit so easily, although I'm no expert on this. The build transforms 13702 modules, most of which are from @mui/icons-material. Even when removing this dependency which lowers the module count to around 3000, I still reach the heap limit. **On 4.1.5 I don't have any issues with the exact same codebase**.

Has there been any change since 4.1 that could have impacted this? I did notice a bump from rollup 3.10 to 3.20, but can't tell from the changelog what the impact is. Locally it builds just fine.

I haven't found anything in https://github.com/vitejs/vite/issues/2433 that helps me.

Last tested on 4.3.9

```
<--- Last few GCs --->

[1249209:0x562bf00]    74516 ms: Scavenge 1908.7 (1986.9) -> 1908.4 (1997.6) MB, 14.6 / 0.0 ms  (average mu = 0.605, current mu = 0.406) allocation failure; 
[1249209:0x562bf00]    74538 ms: Scavenge 1915.1 (1997.6) -> 1915.3 (1997.9) MB, 19.5 / 0.0 ms  (average mu = 0.605, current mu = 0.406) allocation failure; 
[1249209:0x562bf00]    75381 ms: Scavenge 1916.1 (1997.9) -> 1915.2 (2020.1) MB, 843.1 / 0.0 ms  (average mu = 0.605, current mu = 0.406) allocation failure; 


<--- JS stacktrace --->

FATAL ERROR: Reached heap limit Allocation failed - JavaScript heap out of memory
(Omitted some lines for brevity)
Aborted (core dumped)
error Command failed with exit code 1[34]
Error: Process completed with exit code 134.
```

### Reproduction

Not sure how to repro this without sharing my code

### Steps to reproduce

`vite build --mode testing --sourcemap false` on the CI server.

### System Info

```shell
System:
    OS: Linux 5.15 Ubuntu 22.04.2 LTS 22.04.2 LTS (Jammy Jellyfish)
    CPU: (2) x64 DO-Premium-AMD
    Memory: 2.79 GB / 3.83 GB
    Container: Yes
    Shell: 5.1.16 - /bin/bash
  Binaries:
    Node: 12.22.9 - /usr/bin/node
    Yarn: 1.22.19 - /usr/local/bin/yarn
    npm: 8.5.1 - /usr/bin/npm
```


### Used Package Manager

yarn

### Logs

<details>
<summary>Click to expand!</summary>

```shell
2023-06-26T10:01:30.915Z vite:config bundled config file loaded in 636.75ms
2023-06-26T10:01:30.990Z vite:esbuild 63.34ms tsconfck init /home/github/my_path_here
2023-06-26T10:01:31.012Z vite:config using resolved config: {
  build: {
    target: [ 'es2020', 'edge88', 'firefox78', 'chrome87', 'safari14' ],
    cssTarget: [ 'es2020', 'edge88', 'firefox78', 'chrome87', 'safari14' ],
    outDir: 'dist',
    assetsDir: 'assets',
vite v4.3.9 building for testing...
    assetsInlineLimit: 4096,
    cssCodeSplit: true,
    sourcemap: 'false',
    rollupOptions: { onwarn: [Function: onwarn] },
    minify: 'esbuild',
    terserOptions: {},
    write: true,
    emptyOutDir: null,
    copyPublicDir: true,
    manifest: false,
    lib: false,
    ssr: false,
    ssrManifest: false,
    ssrEmitAssets: false,
    reportCompressedSize: true,
    chunkSizeWarningLimit: 500,
    watch: null,
    commonjsOptions: { include: [Array], extensions: [Array] },
    dynamicImportVarsOptions: { warnOnError: true, exclude: [Array] },
    modulePreload: { polyfill: true },
    cssMinify: true
  },
  plugins: [
    'vite:build-metadata',
    'vite:watch-package-data',
    'vite:pre-alias',
    'alias',
    'vite:react-babel',
    'vite:react-refresh',
    'vite-tsconfig-paths',
    'vite:modulepreload-polyfill',
    'vite:resolve',
    'vite:html-inline-proxy',
    'vite:css',
    'vite:esbuild',
    'vite:json',
    'vite:wasm-helper',
    'vite:worker',
    'vite:asset',
    'vite-plugin-svgr',
    'replace',
    'sentry-noop-plugin',
    'vite:wasm-fallback',
    'vite:define',
    'vite:css-post',
    'vite:build-html',
    'vite:worker-import-meta-url',
    'vite:asset-import-meta-url',
    'vite:force-systemjs-wrap-complete',
    'commonjs',
    'vite:data-uri',
    'vite:dynamic-import-vars',
    'vite:import-glob',
    'vite:build-import-analysis',
    'vite:esbuild-transpile',
    'vite:terser',
    'vite:reporter',
    'vite:load-fallback'
  ],
  server: {
    preTransformRequests: true,
    host: 'my.host.here',
    open: true,
    port: 8000,
    sourcemapIgnoreList: [Function: isInNodeModules],
    middlewareMode: false,
    fs: { strict: true, allow: [Array], deny: [Array] }
  },
  mode: 'testing',
  optimizeDeps: {
    disabled: 'build',
    force: undefined,
    include: [ 'react', 'react/jsx-dev-runtime' ],
    esbuildOptions: { preserveSymlinks: false }
  },
  esbuild: { jsxDev: false, jsx: 'automatic', jsxImportSource: undefined },
  resolve: {
    mainFields: [ 'module', 'jsnext:main', 'jsnext' ],
    browserField: true,
    conditions: [],
    extensions: [
      '.mjs',  '.js',
      '.mts',  '.ts',
      '.jsx',  '.tsx',
      '.json'
    ],
    dedupe: [ 'react', 'react-dom' ],
    preserveSymlinks: false,
    alias: [ [Object], [Object] ]
  },
  configFile: '/home/github/my_path_here/vite.config.ts',
  configFileDependencies: [
    '/home/github/my_path_here/vite.config.ts'
  ],
  inlineConfig: {
    root: undefined,
    base: undefined,
    mode: 'testing',
    configFile: undefined,
    logLevel: undefined,
    clearScreen: undefined,
    optimizeDeps: { force: undefined },
    build: { sourcemap: 'false' }
  },
  root: '/home/github/my_path_here',
  base: '/',
  rawBase: '/',
  publicDir: '/home/github/my_path_here/public',
  cacheDir: '/home/github/my_path_here/node_modules/.vite',
  command: 'build',
  ssr: {
    format: 'esm',
    target: 'node',
    optimizeDeps: { disabled: true, esbuildOptions: [Object] }
  },
  isWorker: false,
  mainConfig: null,
  isProduction: true,
  preview: {
    port: undefined,
    strictPort: undefined,
    host: 'my.host.here',
    https: undefined,
    open: true,
    proxy: undefined,
    cors: undefined,
    headers: undefined
  },
  envDir: '/home/github/my_path_here',
  env: {
    VITE_APP_API_URL: 'http://my.url.here/api',
    BASE_URL: '/',
    MODE: 'testing',
    DEV: false,
    PROD: true
  },
  assetsInclude: [Function: assetsInclude],
  logger: {
    hasWarned: false,
    info: [Function: info],
    warn: [Function: warn],
    warnOnce: [Function: warnOnce],
    error: [Function: error],
    clearScreen: [Function: clearScreen],
    hasErrorLogged: [Function: hasErrorLogged]
  },
  packageCache: Map(1) {
    'fnpd_/home/github/my_path_here' => {
      dir: '/home/github/my_path_here',
      data: [Object],
      hasSideEffects: [Function: hasSideEffects],
      webResolvedImports: {},
      nodeResolvedImports: {},
      setResolvedCache: [Function: setResolvedCache],
      getResolvedCache: [Function: getResolvedCache]
    },
    set: [Function (anonymous)]
  },
  createResolver: [Function: createResolver],
  worker: {
    format: 'iife',
    plugins: [
      'vite:build-metadata',
      'vite:watch-package-data',
      'vite:pre-alias',
      'alias',
      'vite:modulepreload-polyfill',
      'vite:resolve',
      'vite:html-inline-proxy',
      'vite:css',
      'vite:esbuild',
      'vite:json',
      'vite:wasm-helper',
      'vite:worker',
      'vite:asset',
      'vite:wasm-fallback',
      'vite:define',
      'vite:css-post',
      'vite:build-html',
      'vite:worker-import-meta-url',
      'vite:asset-import-meta-url',
      'vite:force-systemjs-wrap-complete',
      'commonjs',
      'vite:data-uri',
      'vite:worker-post',
      'vite:dynamic-import-vars',
      'vite:import-glob',
      'vite:build-import-analysis',
      'vite:esbuild-transpile',
      'vite:terser',
      'vite:load-fallback'
    ],
    rollupOptions: {},
    getSortedPlugins: [Function: getSortedPlugins],
    getSortedPluginHooks: [Function: getSortedPluginHooks]
  },
  appType: 'spa',
  experimental: { importGlobRestoreExtension: false, hmrPartialAccept: false },
  getSortedPlugins: [Function: getSortedPlugins],
  getSortedPluginHooks: [Function: getSortedPluginHooks]
}
transforming...
2023-06-26T10:01:32.140Z vite:resolve 2.30ms src/styles/variables -> null
2023-06-26T10:01:32.179Z vite:resolve 8.30ms src/styles/mixins -> null
2023-06-26T10:01:35.065Z vite:resolve 5.75ms ../assets/fonts/Campton-Thin.woff -> /home/github/my_path_here/src/assets/fonts/Campton-Thin.woff
2023-06-26T10:01:35.066Z vite:resolve 6.06ms ../assets/fonts/Campton-ThinItalic.woff -> /home/github/my_path_here/src/assets/fonts/Campton-ThinItalic.woff
2023-06-26T10:01:35.066Z vite:resolve 6.16ms ../assets/fonts/Campton-ExtraLight.woff -> /home/github/my_path_here/src/assets/fonts/Campton-ExtraLight.woff
2023-06-26T10:01:35.066Z vite:resolve 6.21ms ../assets/fonts/Campton-ExtraLightItalic.woff -> /home/github/my_path_here/src/assets/fonts/Campton-ExtraLightItalic.woff
2023-06-26T10:01:35.069Z vite:resolve 6.25ms ../assets/fonts/Campton-Light.woff -> /home/github/my_path_here/src/assets/fonts/Campton-Light.woff
2023-06-26T10:01:35.070Z vite:resolve 9.90ms ../assets/fonts/Campton-LightItalic.woff -> /home/github/my_path_here/src/assets/fonts/Campton-LightItalic.woff
2023-06-26T10:01:35.070Z vite:resolve 10.43ms ../assets/fonts/Campton-Book.woff -> /home/github/my_path_here/src/assets/fonts/Campton-Book.woff
2023-06-26T10:01:35.071Z vite:resolve 10.87ms ../assets/fonts/Campton-BookItalic.woff -> /home/github/my_path_here/src/assets/fonts/Campton-BookItalic.woff
2023-06-26T10:01:35.071Z vite:resolve 11.45ms ../assets/fonts/Campton-Medium.woff -> /home/github/my_path_here/src/assets/fonts/Campton-Medium.woff
2023-06-26T10:01:35.072Z vite:resolve 11.76ms ../assets/fonts/Campton-MediumItalic.woff -> /home/github/my_path_here/src/assets/fonts/Campton-MediumItalic.woff
2023-06-26T10:01:35.072Z vite:resolve 11.87ms ../assets/fonts/Campton-SemiBold.woff -> /home/github/my_path_here/src/assets/fonts/Campton-SemiBold.woff
2023-06-26T10:01:35.072Z vite:resolve 12.17ms ../assets/fonts/Campton-SemiBoldItalic.woff -> /home/github/my_path_here/src/assets/fonts/Campton-SemiBoldItalic.woff
2023-06-26T10:01:35.073Z vite:resolve 12.44ms ../assets/fonts/Campton-Bold.woff -> /home/github/my_path_here/src/assets/fonts/Campton-Bold.woff
2023-06-26T10:01:35.073Z vite:resolve 12.79ms ../assets/fonts/Campton-BoldItalic.woff -> /home/github/my_path_here/src/assets/fonts/Campton-BoldItalic.woff
2023-06-26T10:01:35.074Z vite:resolve 13.04ms ../assets/fonts/Campton-ExtraBold.woff -> /home/github/my_path_here/src/assets/fonts/Campton-ExtraBold.woff
2023-06-26T10:01:35.074Z vite:resolve 13.32ms ../assets/fonts/Campton-ExtraBoldItalic.woff -> /home/github/my_path_here/src/assets/fonts/Campton-ExtraBoldItalic.woff
2023-06-26T10:01:35.074Z vite:resolve 13.65ms ../assets/fonts/Campton-Black.woff -> /home/github/my_path_here/src/assets/fonts/Campton-Black.woff
2023-06-26T10:01:35.075Z vite:resolve 14.00ms ../assets/fonts/Campton-BlackItalic.woff -> /home/github/my_path_here/src/assets/fonts/Campton-BlackItalic.woff
2023-06-26T10:01:35.374Z vite:resolve 0.35ms ../assets/fonts/Campton-ExtraLight.otf -> /home/github/my_path_here/src/assets/fonts/Campton-ExtraLight.otf
2023-06-26T10:01:35.375Z vite:resolve 0.26ms ../assets/fonts/Campton-ExtraLightItalic.otf -> /home/github/my_path_here/src/assets/fonts/Campton-ExtraLightItalic.otf
2023-06-26T10:01:35.376Z vite:resolve 0.26ms ../assets/fonts/Campton-Thin.otf -> /home/github/my_path_here/src/assets/fonts/Campton-Thin.otf
2023-06-26T10:01:35.377Z vite:resolve 0.31ms ../assets/fonts/Campton-ThinItalic.otf -> /home/github/my_path_here/src/assets/fonts/Campton-ThinItalic.otf
2023-06-26T10:01:35.378Z vite:resolve 0.27ms ../assets/fonts/Campton-Light.otf -> /home/github/my_path_here/src/assets/fonts/Campton-Light.otf
2023-06-26T10:01:35.379Z vite:resolve 0.30ms ../assets/fonts/Campton-Book.otf -> /home/github/my_path_here/src/assets/fonts/Campton-Book.otf
2023-06-26T10:01:35.380Z vite:resolve 0.39ms ../assets/fonts/Campton-BookItalic.otf -> /home/github/my_path_here/src/assets/fonts/Campton-BookItalic.otf
2023-06-26T10:01:35.381Z vite:resolve 0.26ms ../assets/fonts/Campton-LightItalic.otf -> /home/github/my_path_here/src/assets/fonts/Campton-LightItalic.otf
2023-06-26T10:01:35.382Z vite:resolve 0.26ms ../assets/fonts/Campton-Medium.otf -> /home/github/my_path_here/src/assets/fonts/Campton-Medium.otf
2023-06-26T10:01:35.383Z vite:resolve 0.30ms ../assets/fonts/Campton-MediumItalic.otf -> /home/github/my_path_here/src/assets/fonts/Campton-MediumItalic.otf
2023-06-26T10:01:35.384Z vite:resolve 0.21ms ../assets/fonts/Campton-SemiBold.otf -> /home/github/my_path_here/src/assets/fonts/Campton-SemiBold.otf
2023-06-26T10:01:35.385Z vite:resolve 0.27ms ../assets/fonts/Campton-SemiBoldItalic.otf -> /home/github/my_path_here/src/assets/fonts/Campton-SemiBoldItalic.otf
2023-06-26T10:01:35.386Z vite:resolve 0.43ms ../assets/fonts/Campton-Bold.otf -> /home/github/my_path_here/src/assets/fonts/Campton-Bold.otf
2023-06-26T10:01:35.388Z vite:resolve 0.28ms ../assets/fonts/Campton-BoldItalic.otf -> /home/github/my_path_here/src/assets/fonts/Campton-BoldItalic.otf
2023-06-26T10:01:35.393Z vite:resolve 0.55ms ../assets/fonts/Campton-ExtraBold.otf -> /home/github/my_path_here/src/assets/fonts/Campton-ExtraBold.otf
2023-06-26T10:01:35.609Z vite:resolve 0.66ms ../assets/fonts/Campton-ExtraBoldItalic.otf -> /home/github/my_path_here/src/assets/fonts/Campton-ExtraBoldItalic.otf
2023-06-26T10:01:35.610Z vite:resolve 0.34ms ../assets/fonts/Campton-BlackItalic.otf -> /home/github/my_path_here/src/assets/fonts/Campton-BlackItalic.otf
2023-06-26T10:01:35.611Z vite:resolve 0.41ms ../assets/fonts/Campton-Black.otf -> /home/github/my_path_here/src/assets/fonts/Campton-Black.otf
2023-06-26T10:01:41.105Z vite:resolve 1.19ms src/styles/_variables.scss -> null
2023-06-26T10:01:41.134Z vite:resolve 18.07ms src/styles/_mixins.scss -> null
✓ 13702 modules transformed.
```
</details>

### Validations

- [X] Follow our [Code of Conduct](https://github.com/vitejs/vite/blob/main/CODE_OF_CONDUCT.md)
- [X] Read the [Contributing Guidelines](https://github.com/vitejs/vite/blob/main/CONTRIBUTING.md).
- [X] Read the [docs](https://vitejs.dev/guide).
- [X] Check that there isn't [already an issue](https://github.com/vitejs/vite/issues) that reports the same bug to avoid creating a duplicate.
- [X] Make sure this is a Vite issue and not a framework-specific issue. For example, if it's a Vue SFC related bug, it should likely be reported to [vuejs/core](https://github.com/vuejs/core) instead.
- [X] Check that this is a concrete bug. For Q&A open a [GitHub Discussion](https://github.com/vitejs/vite/discussions) or join our [Discord Chat Server](https://chat.vitejs.dev/).
- [X] The provided reproduction is a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) of the bug.

--- a/packages/vite/src/node/cli.ts
+++ b/packages/vite/src/node/cli.ts
@@ -70,9 +70,9 @@ const filterDuplicateOptions = <T extends object>(options: T) => {
 /**
  * removing global flags before passing as command specific sub-configs
  */
-function cleanOptions<Options extends GlobalCLIOptions>(
-  options: Options,
-): Omit<Options, keyof GlobalCLIOptions> {
+function cleanOptions<
+  Options extends GlobalCLIOptions & (ServerOptions | BuildOptions),
+>(options: Options): Omit<Options, keyof GlobalCLIOptions> {
   const ret = { ...options }
   delete ret['--']
   delete ret.c
@@ -87,6 +87,18 @@ function cleanOptions<Options extends GlobalCLIOptions>(
   delete ret.filter
   delete ret.m
   delete ret.mode
+
+  // convert the sourcemap option to a boolean if necessary
+  if ('sourcemap' in ret) {
+    const sourcemap = ret.sourcemap as `${boolean}` | 'inline' | 'hidden'
+    ret.sourcemap =
+      sourcemap === 'true'
+        ? true
+        : sourcemap === 'false'
+        ? false
+        : ret.sourcemap
+  }
+
   return ret
 }
 

--- a/packages/vite/src/node/cli.ts
+++ b/packages/vite/src/node/cli.ts
@@ -70,9 +70,9 @@ const filterDuplicateOptions = <T extends object>(options: T) => {
 /**
  * removing global flags before passing as command specific sub-configs
  */
-function cleanOptions<
-  Options extends GlobalCLIOptions & (ServerOptions | BuildOptions),
->(options: Options): Omit<Options, keyof GlobalCLIOptions> {
+function cleanOptions<Options extends GlobalCLIOptions>(
+  options: Options,
+): Omit<Options, keyof GlobalCLIOptions> {
   const ret = { ...options }
   delete ret['--']
   delete ret.c

